### PR TITLE
Feature 10861

### PR DESCRIPTION
**Pull request by ashish-greycube**
Please refer design flow chart as below.
image

Animated gif explaining the "Bypass Customer Credit Limit" check feature
and "Customer Credit Balance Report" in action.

Outstanding Amount is not updated on submission of "Sales Order".
It is updated during full/partial payment against "Sales Order" and/or after submitting "Delivery Note". (as per the case)
image

Test Case :
Automated test cases passed for both scenarios.

   1. With Credit Limit check Bypass at Sales Order 
    
image

2. With Credit Limit check without bypass at Sales Order (default)
image

### CODE REVIEW

**Comment by ashish-greycube** [comment]
This code check is in context to https://github.com/frappe/erpnext/issues/10861

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 1]
```diff
@@ -196,7 +196,11 @@ def update_project(self):
 
 	def check_credit_limit(self):
 		from erpnext.selling.doctype.customer.customer import check_credit_limit
-		check_credit_limit(self.customer, self.company)
+	    #PR : 10861, Author : ashish-greycube & jigneshpshah,  Email:mr.ashish.shah@gmail.com 
+		# bypass credit limit check is set to true (1) at sales order level, then we need not to check credit limit and vise versa
+		bypass_credit_limit_check_at_sales_order = frappe.db.get_value("Customer", self.customer, "bypass_credit_limit_check_at_sales_order")
```
Comment (tundebabzy) [inline erpnext/selling/doctype/sales_order/sales_order.py]: bypass_credit_limit_check_at_sales_order = cint(frappe.db.........)

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 2]
```diff
@@ -461,6 +465,14 @@ def set_missing_values(source, target):
 				target.po_no = ", ".join(list(set(target_po_no))) if len(target_po_no) > 1 else target_po_no[0]
 			else:
 				target.po_no = source.po_no
+	    
+	    #PR : 10861, Author : ashish-greycube & jigneshpshah,  Email:mr.ashish.shah@gmail.com 
+		# Since the credit limit check is bypassed at sales order level, we need to check it at delivery note
+		bypass_credit_limit_check_at_sales_order = frappe.db.get_value("Customer", source.customer, "bypass_credit_limit_check_at_sales_order")
```
Comment (tundebabzy) [inline erpnext/selling/doctype/sales_order/sales_order.py]: bypass_credit_limit_check_at_sales_order = cint(frappe.db....)

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 3]
```diff
@@ -524,6 +536,13 @@ def set_missing_values(source, target):
 		target.run_method("set_missing_values")
 		target.run_method("calculate_taxes_and_totals")
 
+	    #PR : 10861, Author : ashish-greycube & jigneshpshah,  Email:mr.ashish.shah@gmail.com 
+		# Since the credit limit check is bypassed at sales order level, we need to check it at sales invoice
+		bypass_credit_limit_check_at_sales_order = frappe.db.get_value("Customer", source.customer, "bypass_credit_limit_check_at_sales_order")
```
Comment (tundebabzy) [inline erpnext/selling/doctype/sales_order/sales_order.py]: bypass_credit_limit_check_at_sales_order = cint(frappe.db.....)

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py [hunk 4]
```diff
@@ -70,7 +70,6 @@ def validate(self):
 		self.clear_unallocated_advances("Sales Invoice Advance", "advances")
 		self.add_remarks()
 		self.validate_write_off_account()
-		self.validate_duplicate_offline_pos_entry()
```
Comment (tundebabzy) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.py]: Why did you remove?

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py [hunk 5]
```diff
@@ -463,12 +462,6 @@ def validate_write_off_account(self):
 		if flt(self.write_off_amount) and not self.write_off_account:
 			msgprint(_("Please enter Write Off Account"), raise_exception=1)
 
-	def validate_duplicate_offline_pos_entry(self):
```
Comment (tundebabzy) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.py]: Why did you remove?

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py [hunk 6]
```diff
@@ -924,6 +917,13 @@ def set_missing_values(source, target):
 		target.run_method("set_missing_values")
 		target.run_method("calculate_taxes_and_totals")
 
+	    #PR : 10861, Author : ashish-greycube & jigneshpshah,  Email:mr.ashish.shah@gmail.com 
+		# Since the credit limit check is bypassed at sales order level, we need to check it at delivery note
+		bypass_credit_limit_check_at_sales_order = frappe.db.get_value("Customer", source.customer, "bypass_credit_limit_check_at_sales_order")
```
Comment (tundebabzy) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.py]: bypass_credit_limit_check_at_sales_order = cint(frappe.db....)

File: erpnext/selling/doctype/customer/customer.py [hunk 7]
```diff
@@ -186,6 +188,20 @@ def get_customer_outstanding(customer, company):
 
 	outstanding_based_on_so = flt(outstanding_based_on_so[0][0]) if outstanding_based_on_so else 0.0
 
+	#PR : 10861, Author : ashish-greycube & jigneshpshah,  Email:mr.ashish.shah@gmail.com 
+	# Since the credit limit check is bypassed at sales order level, when customer credit balance report is run we need to treat sales order with status 'To Deliver and Bill' as not outstanding
+	outstanding_based_on_bypassed_so = 0.0
+	bypass_credit_limit_check_at_sales_order = frappe.db.get_value("Customer", customer, "bypass_credit_limit_check_at_sales_order")
```
Comment (tundebabzy) [inline erpnext/selling/doctype/customer/customer.py]: bypass_credit_limit_check_at_sales_order = cint(frappe.db....)

File: erpnext/selling/doctype/customer/test_customer.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
+/* eslint-disable */
```
Comment (tundebabzy) [inline erpnext/selling/doctype/customer/test_customer.js]: This looks like the default template. How about update it or remove it completely.

**Review by tundebabzy — CHANGES_REQUESTED** [review]

File: erpnext/selling/doctype/customer/customer.json [hunk 9]
```diff
@@ -911,6 +911,36 @@
...
+   "print_hide": 0, 
+   "print_hide_if_no_value": 0, 
+   "read_only": 0, 
+   "remember_last_selected_value": 0, 
+   "report_hide": 0, 
+   "reqd": 0, 
+   "search_index": 0, 
+   "set_only_once": 0, 
+   "unique": 0
+  },
```
Comment (tundebabzy) [inline erpnext/selling/doctype/customer/customer.json]: Explicitly set default value for `bypass_credit_limit_check_at_sales_order` to 0 since explicit is always better.

**Review by tundebabzy — CHANGES_REQUESTED** [review]

**Comment by tundebabzy** [comment]
- After setting default to 0, you will also need to add a patch to update old records.
- Still missing python tests
- Still missing updates to the manual about this new feature

File: erpnext/selling/report/customer_credit_balance/customer_credit_balance.py [hunk 10]
```diff
@@ -49,6 +52,7 @@ def get_details(filters):
 
 	if filters.get("customer"):
 		conditions += " where name = %(customer)s"
-
-	return frappe.db.sql("""select name, customer_name from `tabCustomer` %s""" 
+	#PR : 10861, Author : ashish-greycube & jigneshpshah,  Email:mr.ashish.shah@gmail.com 
+	# Return column bypass credit limit check at sales order level	
+	return frappe.db.sql("""select name, customer_name,bypass_credit_limit_check_at_sales_order from `tabCustomer` %s"""
```
Comment (tundebabzy) [inline erpnext/selling/report/customer_credit_balance/customer_credit_balance.py]: This would cause an SQL Injection vulnerability. Please correct this

**Review by tundebabzy — CHANGES_REQUESTED** [review]

**Comment by ashish-greycube** [comment]
Hi @tundebabzy , please find the update as requested
1)All the comments are incorporated
2)Still missing updates to the manual about this new feature. Ashish: Documentation changes pull request is [11343](https://github.com/frappe/erpnext/pull/11343)
3)customer_credit_balance.py :This would cause an SQL Injection vulnerability. Please correct this. Ashish: I have just added a new column to select query. The structure of SQL is as it is and not changed by me.
4)Still missing python tests. Ashish : JS test case are written which covers the new flow. Also regression test case is added. This are created using [guide](https://frappe.io/docs/user/en/guides/automated-testing/qunit-testing) . Test result are also shared. Test case are self sufficient i.e. It creates the requierd test data.

**Comment by ashish-greycube** [comment]
by mistake i closed it

**Comment by ashish-greycube** [comment]
@tundebabzy please let me know if it is good to go!

**Comment by ashish-greycube** [comment]
@nabinhait , It appears the pull is assigned to you. Please guide. thanks

**Comment by ashish-greycube** [comment]
hi @nabinhait  @tundebabzy , I have meet all the requirements or changes suggested. Please guide. Also note that in this case javascript test case will only make sense as when submit button is clicked from UI the new business logic code gets executed, whereas python test case of sales order submits directly.
So, please review and help me to merge my first pull request. thanks!

**Comment by ashish-greycube** [comment]
Hi @nabinhait request you to help me merge the PR. thanks!

File: erpnext/patches/v9_0/set_bypasscreditlimitcheckatsalesorder_in_customer.py [hunk 11]
```diff
@@ -0,0 +1,10 @@
+import frappe
+
+def execute():
+	frappe.reload_doctype("Customer")
+	
+	if "bypass_credit_limit_check_at_sales_order" in frappe.db.get_table_columns("Customer"):
+		frappe.db.sql("""
+			update `tabCustomer`
+			set bypass_credit_limit_check_at_sales_order = 0
```
Comment (nabinhait) [inline erpnext/patches/v9_0/set_bypasscreditlimitcheckatsalesorder_in_customer.py]: I think the patch is not needed, system should set "0" by default.

File: erpnext/selling/doctype/customer/customer.py [hunk 12]
```diff
@@ -186,6 +188,20 @@ def get_customer_outstanding(customer, company):
 
 	outstanding_based_on_so = flt(outstanding_based_on_so[0][0]) if outstanding_based_on_so else 0.0
 
+	#PR : 10861, Author : ashish-greycube & jigneshpshah,  Email:mr.ashish.shah@gmail.com 
+	# Since the credit limit check is bypassed at sales order level, when customer credit balance report is run we need to treat sales order with status 'To Deliver and Bill' as not outstanding
+	outstanding_based_on_bypassed_so = 0.0
+	bypass_credit_limit_check_at_sales_order = cint(frappe.db.get_value("Customer", customer, "bypass_credit_limit_check_at_sales_order"))
+	if bypass_credit_limit_check_at_sales_order == 1 and caller=='customer_credit_balance_report':
+		outstanding_based_on_bypassed_so = frappe.db.sql("""
```
Comment (nabinhait) [inline erpnext/selling/doctype/customer/customer.py]: Why this query needed?
We can skip `outstanding_based_on_so` optionally, like in case of report.

File: erpnext/selling/doctype/customer/customer.py [hunk 13]
```diff
@@ -170,7 +170,9 @@ def check_credit_limit(customer, company):
 			throw(_("Please contact to the user who have Sales Master Manager {0} role")
 				.format(" / " + credit_controller if credit_controller else ""))
 
-def get_customer_outstanding(customer, company):
+#PR : 10861, Author : ashish-greycube & jigneshpshah,  Email:mr.ashish.shah@gmail.com 
+# get_customer_outstanding is a very generic function which is invoked from many places. A third non mandatory argument is added to change its behaviour based on caller .
+def get_customer_outstanding(customer, company,caller=None):
```
Comment (nabinhait) [inline erpnext/selling/doctype/customer/customer.py]: The report name should not be hard-coded like this using caller. Tomorrow, if we need to call this function for another purpose, will not be able to do that. It should be a flag, based on purpose.

Comment (nabinhait) [reply erpnext/selling/doctype/customer/customer.py]: Always use space after comma

**Review by nabinhait — CHANGES_REQUESTED** [review]

**Comment by nabinhait** [comment]
Closing it for now, please reopen after fixing.

File: erpnext/selling/doctype/customer/customer.py (hunk 12 above)
Comment (ashish-greycube) [reply erpnext/selling/doctype/customer/customer.py]: Nabin, there are two things (1) outstanding_based_on_so and (2) outstanding_based_on_bypassed_so [where credit limit is overruled].
When customer_credit_balance_report is run we should return correct outstanding amount by checking each customer for flag "by pass customer credit limit".
We need to treat sales order with status 'To Deliver and Bill'  and where no advance is paid as NOT outstanding
Hence the logic is "outstanding for by pass credit limit" = "outstanding_based_on_so" ---- (minus) "outstanding_based_on_bypassed_so"

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py [hunk 14]
```diff
@@ -924,6 +924,13 @@ def set_missing_values(source, target):
 		target.run_method("set_missing_values")
 		target.run_method("calculate_taxes_and_totals")
 
+	    #PR : 10861, Author : ashish-greycube & jigneshpshah,  Email:mr.ashish.shah@gmail.com 
+		# Since the credit limit check is bypassed at sales order level, we need to check it at delivery note
+		bypass_credit_limit_check_at_sales_order = cint(frappe.db.get_value("Customer", source.customer, "bypass_credit_limit_check_at_sales_order"))
+		if bypass_credit_limit_check_at_sales_order == 1:
+			from erpnext.selling.doctype.customer.customer import check_credit_limit
+			check_credit_limit(source.customer, source.company)
```
[the remark this answers is not in the document]
Comment (ashish-greycube) [reply erpnext/accounts/doctype/sales_invoice/sales_invoice.py]: Agreed.It is removed

File: erpnext/patches.txt [hunk 15]
```diff
@@ -455,3 +455,4 @@ erpnext.patches.v9_0.add_healthcare_domain
 erpnext.patches.v9_0.set_variant_item_description
 erpnext.patches.v9_0.set_uoms_in_variant_field
 erpnext.patches.v9_0.copy_old_fees_field_data
+erpnext.patches.v9_0.set_bypasscreditlimitcheckatsalesorder_in_customer
```
[the remark this answers is not in the document]
Comment (ashish-greycube) [reply erpnext/patches.txt]: As the patch is removed. File is no longer required
